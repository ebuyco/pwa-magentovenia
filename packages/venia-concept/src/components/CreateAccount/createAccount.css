.root {
    display: grid;
    gap: 0.5rem;
    justify-items: stretch;
    padding: 0 1.5rem;
}

.actions {
    display: grid;
    justify-items: center;
    margin-top: 1rem;
    padding: 1rem 0;
}

.lead {
    margin: 1rem 0;
    text-align: center;
}
