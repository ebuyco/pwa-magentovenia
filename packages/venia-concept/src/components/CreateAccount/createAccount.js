import React, { Component } from 'react';
import { func, shape, string } from 'prop-types';
import { Form } from 'informed';

import classify from 'src/classify';
import Button from 'src/components/Button';
import ErrorDisplay from 'src/components/ErrorDisplay';
import Field from 'src/components/Field';
import TextInput from 'src/components/TextInput';
import { asyncValidators, validators } from './validators';
import defaultClasses from './createAccount.css';

class CreateAccount extends Component {
    static propTypes = {
        classes: shape({
            error: string,
            root: string
        }),
        createAccountError: shape({
            message: string
        }),
        initialValues: shape({
            email: string,
            firstName: string,
            lastName: string
        }),
        onSubmit: func
    };

    static defaultProps = {
        initialValues: {}
    };

    get initialValues() {
        const { initialValues } = this.props;
        const { email, firstName, lastName, ...rest } = initialValues;

        return {
            customer: { email, firstname: firstName, lastname: lastName },
            ...rest
        };
    }

    handleSubmit = values => {
        const { onSubmit } = this.props;

        if (typeof onSubmit === 'function') {
            onSubmit(values);
        }
    };

    render() {
        const { handleSubmit, initialValues, props } = this;
        const { classes, createAccountError } = props;

        return (
            <Form
                className={classes.root}
                initialValues={initialValues}
                onSubmit={handleSubmit}
            >
                <h3 className={classes.lead}>
                    {'An account gives you access to rewards!'}
                </h3>
                <Field label="Email">
                    <TextInput
                        field="customer.email"
                        autoComplete="email"
                        validate={validators.get('email')}
                        asyncValidate={asyncValidators.get('email')}
                        validateOnBlur
                        asyncValidateOnBlur
                    />
                </Field>
                <Field label="First Name">
                    <TextInput
                        field="customer.firstname"
                        autoComplete="given-name"
                        validate={validators.get('firstName')}
                        validateOnBlur
                    />
                </Field>
                <Field label="Last Name">
                    <TextInput
                        field="customer.lastname"
                        autoComplete="family-name"
                        validate={validators.get('lastName')}
                        validateOnBlur
                    />
                </Field>
                <Field label="Password">
                    <TextInput
                        field="password"
                        type="password"
                        autoComplete="new-password"
                        validate={validators.get('password')}
                        validateOnBlur
                    />
                </Field>
                <Field label="Confirm Password">
                    <TextInput
                        field="confirm"
                        type="password"
                        validate={validators.get('confirm')}
                        validateOnBlur
                    />
                </Field>
                <ErrorDisplay error={createAccountError} />
                <div className={classes.actions}>
                    <Button type="submit" priority="high">
                        {'Submit'}
                    </Button>
                </div>
            </Form>
        );
    }
}

export default classify(defaultClasses)(CreateAccount);
