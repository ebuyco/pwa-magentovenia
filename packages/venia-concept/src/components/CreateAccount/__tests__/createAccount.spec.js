import React from 'react';
import waitForExpect from 'wait-for-expect';
import TestRenderer from 'react-test-renderer';
import { Form } from 'informed';

import CreateAccount from '../createAccount';
import { asyncValidators, validators } from '../validators';

jest.mock('../validators');

const submitCallback = jest.fn();

afterEach(() => {
    jest.clearAllMocks();
});

test('renders the correct tree', () => {
    const tree = TestRenderer.create(<CreateAccount />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('has a submit handler', () => {
    const { root } = TestRenderer.create(<CreateAccount />);

    const { instance } = root.children[0];

    expect(instance.handleSubmit).toBeInstanceOf(Function);
});

test('attaches the submit handler', () => {
    const { root } = TestRenderer.create(<CreateAccount />);

    const { instance } = root.children[0];
    const { onSubmit } = root.findByType(Form).props;

    expect(onSubmit).toBe(instance.handleSubmit);
});

test('executes validators on submit', async () => {
    const { root } = TestRenderer.create(<CreateAccount />);

    const form = root.findByType(Form);
    const { api } = form.instance.controller;

    // touch fields, call validators, call onSubmit
    api.submitForm();
    // await async validation
    await waitForExpect(() => {
        for (const validator of validators.values()) {
            expect(validator).toHaveBeenCalledTimes(1);
        }
        for (const validator of asyncValidators.values()) {
            expect(validator).toHaveBeenCalledTimes(1);
        }
    });
});

test('calls onSubmit if validation passes', async () => {
    const { root } = TestRenderer.create(
        <CreateAccount onSubmit={submitCallback} />
    );

    const form = root.findByType(Form);
    const { api } = form.instance.controller;

    // touch fields, call validators, call onSubmit
    api.submitForm();
    // await async validation
    await waitForExpect(() => {
        const { asyncErrors, errors } = form.instance.controller.state;

        expect(Object.keys(asyncErrors)).toHaveLength(0);
        expect(Object.keys(errors)).toHaveLength(0);
        expect(submitCallback).toHaveBeenCalledTimes(1);
    });
});
