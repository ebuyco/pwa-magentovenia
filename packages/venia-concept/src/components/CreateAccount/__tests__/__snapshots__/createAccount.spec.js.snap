// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct tree 1`] = `
<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <h3>
    An account gives you access to rewards!
  </h3>
  <div>
    <span>
      Email
    </span>
    <span
      style={
        Object {
          "--iconsAfter": 0,
          "--iconsBefore": 0,
        }
      }
    >
      <span>
        <input
          autoComplete="email"
          name="customer.email"
          onBlur={[Function]}
          onChange={[Function]}
          value=""
        />
      </span>
      <span />
      <span />
    </span>
    <p />
  </div>
  <div>
    <span>
      First Name
    </span>
    <span
      style={
        Object {
          "--iconsAfter": 0,
          "--iconsBefore": 0,
        }
      }
    >
      <span>
        <input
          autoComplete="given-name"
          name="customer.firstname"
          onBlur={[Function]}
          onChange={[Function]}
          value=""
        />
      </span>
      <span />
      <span />
    </span>
    <p />
  </div>
  <div>
    <span>
      Last Name
    </span>
    <span
      style={
        Object {
          "--iconsAfter": 0,
          "--iconsBefore": 0,
        }
      }
    >
      <span>
        <input
          autoComplete="family-name"
          name="customer.lastname"
          onBlur={[Function]}
          onChange={[Function]}
          value=""
        />
      </span>
      <span />
      <span />
    </span>
    <p />
  </div>
  <div>
    <span>
      Password
    </span>
    <span
      style={
        Object {
          "--iconsAfter": 0,
          "--iconsBefore": 0,
        }
      }
    >
      <span>
        <input
          autoComplete="new-password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </span>
      <span />
      <span />
    </span>
    <p />
  </div>
  <div>
    <span>
      Confirm Password
    </span>
    <span
      style={
        Object {
          "--iconsAfter": 0,
          "--iconsBefore": 0,
        }
      }
    >
      <span>
        <input
          name="confirm"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </span>
      <span />
      <span />
    </span>
    <p />
  </div>
  <div>
    <button
      type="submit"
    >
      <span>
        Submit
      </span>
    </button>
  </div>
</form>
`;
